File: src/test/generate_test_data/generateRunSaveTest_constructionModelesLassoRank.R
source("helpers/constructionModelesLassoRank.R")

generateRunSaveTest_constructionModelesLassoRank = function(n=200, p=15, m=10, k=3, L=12, mini=5,
	maxi=10, gamma=1.0, rangmin=3, rangmax=6)
{
  tau = 1e-6
  pi = matrix(1./k, nrow=k, ncol=L)
  rho = array(dim=c(m,m,k,L))
  for (l in 1:L)
	{
    for (r in 1:k)
      rho[,,r,l] = diag(1,m)
  }
  A1 = matrix(seq_len(p), nrow=p, ncol=L)
	require(valse)
  xy = valse:::generateXYdefault(n, p, m, k)

  testFolder = "../data/"
  dir.create(testFolder, showWarnings=FALSE, mode="0755")
  #save inputs
  write.table(as.double(pi), paste(testFolder,"pi",sep=""),
		row.names=F, col.names=F)
  write.table(as.double(rho),paste(testFolder,"rho",sep=""),
		row.names=F, col.names=F)
  write.table(as.integer(mini), paste(testFolder,"mini",sep=""),
		row.names=F, col.names=F)
  write.table(as.integer(maxi), paste(testFolder,"maxi",sep=""),
		row.names=F, col.names=F)
  write.table(as.double(xy$X), paste(testFolder,"X",sep=""),
		row.names=F, col.names=F)
  write.table(as.double(xy$Y), paste(testFolder,"Y",sep=""),
		row.names=F, col.names=F)
  write.table(as.double(tau),paste(testFolder,"tau",sep=""),
		row.names=F, col.names=F)
  write.table(as.double(A1),paste(testFolder,"A1",sep=""),
		row.names=F, col.names=F)
  write.table(as.integer(rangmin),paste(testFolder,"rangmin",sep=""),
		row.names=F, col.names=F)
  write.table(as.integer(rangmax),paste(testFolder,"rangmax",sep=""),
		row.names=F, col.names=F)
  write.table(as.integer(c(n,p,m,k,L)),paste(testFolder,"dimensions",sep=""),
		row.names=F, col.names=F)

  res = constructionModelesLassoRank(pi,rho,mini,maxi,xy$X,xy$Y,tau,A1,rangmin,rangmax)

  #save output
  write.table(as.double(res$phi), paste(testFolder,"phi",sep=""), row.names=F, col.names=F)
  write.table(as.double(res$llh), paste(testFolder,"llh",sep=""), row.names=F, col.names=F)
}

File: src/test/generate_test_data/generateRunSaveTest_constructionModelesLassoRank.m
function[] = generateRunSaveTest_constructionModelesLassoRank(n, p, m, k, L, mini, maxi, gamma, rangmin, rangmax, varargin)

	%set defaults for optional inputs
	optargs = {200 15 10 3 12 5 10 1.0 3 6};
	%replace defaults by user parameters
	optargs(1:length(varargin)) = varargin;
	[n, p, m, k, L, mini, maxi, gamma, rangmin, rangmax] = optargs{:};
	mini = int64(mini);
	maxi = int64(maxi);
	rangmin = int64(rangmin);
	rangmax = int64(rangmax);
	tau = 1e-6;

	Pi = zeros(k,L);
	for l=1:L
		Pi(:,l) = (1.0/k)*ones(1,k);
	end
	Rho = zeros(m,m,k,L);
	for l=1:L
		for r=1:k
			Rho(:,:,r,l) = eye(m);
		end
	end

	%Generate X and Y
	[X, Y, ~] = generateIOdefault(n, p, m, k);

	A1 = zeros(p,L,'int64');
    for i=1:L
        A1(:,i) = 1:p;
    end

	%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

	testFolder = 'data/';
	mkdir(testFolder);
	delimiter = ' ';

	%save inputs
	dlmwrite(strcat(testFolder,'Rho'), reshape(Rho,1,[]), delimiter);
	dlmwrite(strcat(testFolder,'Pi'), reshape(Pi,1,[]), delimiter);
	dlmwrite(strcat(testFolder,'mini'), mini, delimiter);
	dlmwrite(strcat(testFolder,'maxi'), maxi, delimiter);
	dlmwrite(strcat(testFolder,'X'), reshape(X,1,[]), delimiter);
	dlmwrite(strcat(testFolder,'Y'), reshape(Y,1,[]), delimiter);
	dlmwrite(strcat(testFolder,'tau'), tau, delimiter);
	dlmwrite(strcat(testFolder,'A1'), reshape(A1,1,[]), delimiter);
	dlmwrite(strcat(testFolder,'rangmin'), rangmin, delimiter);
	dlmwrite(strcat(testFolder,'rangmax'), rangmax, delimiter);
	dlmwrite(strcat(testFolder,'dimensions'), [n,p,m,k,L], delimiter);

	[phi,llh] = constructionModelesLassoRank(Pi,Rho,mini,maxi,X,Y,tau,A1,rangmin,rangmax);

	%save output
	Size = (rangmax-rangmin+1)^k;
	dlmwrite(strcat(testFolder,'phi'), reshape(phi,1,[]), delimiter);
	dlmwrite(strcat(testFolder,'llh'), reshape(llh,1,[]), delimiter);

end
